rest/src/main/java/org/jboss/pnc/rest/provider/ProductMilestoneProvider.java

```diff
@@ -91,4 +122,118 @@ public void removeDistributedArtifact(Integer milestoneId, Integer artifactId) t
         repository.save(milestone);
     }
 
+    @Override
+    public void update(Integer id, ProductMilestoneRest restEntity) throws ValidationException {
+        restEntity.setId(id);
+        validateBeforeUpdating(id, restEntity);
+        ProductMilestone milestone = toDBModel().apply(restEntity);
+
+        if (restEntity.getEndDate() != null && noReleaseInProgress(milestone)) {
+            ProductMilestoneRelease release = startRelease(milestone);
+            releaseRepository.save(release);
+        }
+        repository.save(milestone);
+    }
+
+    private boolean noReleaseInProgress(ProductMilestone milestone) {
+        ProductMilestoneRelease latestRelease = releaseRepository.findLatestByMilestone(milestone);
+
+        return latestRelease == null || latestRelease.getStatus() != BrewPushStatus.IN_PROGRESS;
+    }
+
+    private <T extends BpmNotificationRest> ProductMilestoneRelease startRelease(ProductMilestone milestone) {
+        ProductMilestoneRelease release = new ProductMilestoneRelease();
+        release.setStartingDate(new Date());
+        release.setMilestone(milestone);
+        try {
+            BpmBrewPushTask releaseTask = new BpmBrewPushTask(milestone);
+            releaseTask.addListener(BpmEventType.BREW_IMPORT_SUCCESS, this::onSuccessfulPush);
+            releaseTask.<BpmStringMapNotificationRest>addListener(BpmEventType.BREW_IMPORT_ERROR, r -> onFailedPush(milestone.getId(), r));
+            bpmManager.startTask(releaseTask);
+            release.setLog("Brew push task started\n");
+
+            return release;
+        } catch (CoreException e) {
+            log.error("Error trying to start brew push task for milestone: {}", milestone.getId(), e);
+            release.setLog("Brew push BPM task creation failed. Check log for more details " + e.getMessage() + "\n");
+            release.setStatus(BrewPushStatus.SYSTEM_ERROR);
+            release.setEndDate(new Date());
+            return release;
+        }
+    }
+
+    private void onSuccessfulPush(BrewPushMilestoneResultRest result) {
+        int milestoneId = result.getMilestoneId();
+        ProductMilestone milestone = repository.queryById(milestoneId);
+        String message = describeCompletedPush(result);
+        updateRelease(milestone, message, result.getPushStatus().toBrewPushStatus());
+    }
+
+    private void updateRelease(ProductMilestone milestone, String message, BrewPushStatus status) {
```

Comment:
Indeed, something is wrong with this code being placed here.

What would you say for pulling it out to e.g. `ProductMilestoneReleaseManager`? In the old days it would be called a service, but we don't follow this convention.